src/Moq/Mock.cs

```diff
@@ -390,78 +390,46 @@ internal static void VerifyNoOtherCalls(Mock mock)
 
 		#region Setup
 
-		[DebuggerStepThrough]
-		internal static MethodCall SetupVoid(Mock mock, LambdaExpression expression, Condition condition)
+		internal static MethodCall Setup(Mock mock, LambdaExpression expression, Condition condition)
 		{
-			return PexProtector.Invoke(SetupVoidPexProtected, mock, expression, condition);
-		}
-
-		private static MethodCall SetupVoidPexProtected(Mock mock, LambdaExpression expression, Condition condition)
-		{
-			var (obj, method, args) = expression.GetCallInfo(mock);
-
-			ThrowIfSetupExpressionInvolvesUnsupportedMember(expression, method);
-			ThrowIfSetupMethodNotVisibleToProxyFactory(method);
-			var setup = new MethodCall(mock, condition, expression, method, args);
-
-			var targetMock = GetTargetMock(obj, mock);
-			targetMock.Setups.Add(setup);
+			Guard.NotNull(expression, nameof(expression));
 
-			return setup;
+			return Mock.SetupRecursive(mock, expression, setupLast: (part, targetMock) =>
+			{
+				var setup = new MethodCall(targetMock, condition, part.Expression, part.Method, part.Arguments);
+				targetMock.Setups.Add(setup);
+				return setup;
+			});
 		}
 
-		[DebuggerStepThrough]
-		internal static MethodCall SetupNonVoid(Mock mock, LambdaExpression expression, Condition condition)
+		internal static MethodCall SetupGet(Mock mock, LambdaExpression expression, Condition condition)
 		{
-			return PexProtector.Invoke(SetupNonVoidPexProtected, mock, expression, condition);
-		}
+			Guard.NotNull(expression, nameof(expression));
 
-		private static MethodCall SetupNonVoidPexProtected(Mock mock, LambdaExpression expression, Condition condition)
-		{
-			if (expression.IsProperty())
+			if (!expression.IsPropertyIndexer())  // guard because `.ToPropertyInfo()` doesn't (yet) work for indexers
 			{
-				return SetupGet(mock, expression, condition);
+				var property = expression.ToPropertyInfo();
+				Guard.CanRead(property);
 			}
 
-			var (obj, method, args) = expression.GetCallInfo(mock);
-
-			ThrowIfSetupExpressionInvolvesUnsupportedMember(expression, method);
-			ThrowIfSetupMethodNotVisibleToProxyFactory(method);
-			var setup = new MethodCall(mock, condition, expression, method, args);
-
-			var targetMock = GetTargetMock(obj, mock);
-			targetMock.Setups.Add(setup);
-
-			return setup;
+			return Mock.SetupRecursive(mock, expression, setupLast: (part, targetMock) =>
+			{
+				var setup = new MethodCall(targetMock, condition, part.Expression, part.Method, part.Arguments);
+				targetMock.Setups.Add(setup);
+				return setup;
+			});
```

Comment:
There's some code duplication here, we might just as well delegate to `Setup`, making it clear that the only distinction between `Setup` and `SetupGet` is that the latter performs some additional checks on the setup expression.